packages/server-wallet/src/message-service/types.ts

```diff
@@ -0,0 +1,30 @@
+import {Message} from '@statechannels/client-api-schema';
+
+/**
+ * This is the handler that any messaging service implementation should call when receiving a message
+ * The handler is responsible for pushing the message into the appropriate wallet
+ */
+export type MessageHandler = (
+  // Including the to makes it easier for a messaging service to handle multiple wallets if it chooses to do so
+  to: string,
+  message: unknown,
+  messageService: MessagingServiceInterface
+) => Promise<void>;
+
+/**
+ * A MessageServiceFactory is responsible for generating a MessageService.
+ * Eventually the wallet will require a MessageServiceFactory.
+ * The wallet will use it to construct the messagingService that it can use to send messages.
+ * messageHandler should be triggered whenever the messaging service receives a message.
+ */
+export type MessageServiceFactory = (
+  messageHandler: MessageHandler
+) => Promise<MessagingServiceInterface>;
+
+export interface MessagingServiceInterface {
```

Comment:
Do we have stated standards for declaring interfaces vs. types?

Reply:
I don't think we do!

Generally, I tend to favour an interface when defining something like an API but that's mostly based on habit. 

From the [typescript handbook](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#differences-between-type-aliases-and-interfaces)
```
For the most part, you can choose based on personal preference, and TypeScript will tell you if it needs something to be the other kind of declaration. If you would like a heuristic, use interface until you need to use features from type.
```
so it seems like we should prefer interfaces until they no longer suffice?

Reply:
Can a `Class` implement a `type`?

Reply:
Beat me to it (I also tried it [here](https://www.typescriptlang.org/play?#code/C4TwDgpgBAyglgWzAGwgFXNAvFA3gQwC4A7AVwQCMIAnAXwCh6IAPMAe2uCgGNl8BnfrEQoIAYT6CoI1AgjFgQ+ElQZIeelC1QwpCsjjco+KDgCM9WkA))

Reply:
The differences between interfaces and types are small. The features from type that might apply in the above heuristic are in creating and extending from union types. Interfaces support declaration merging which can sometimes matter for user-augmentation of a library's exported types. Anyway, the usage here is not a problem at all, I just wanted to ask because I saw both types and an interface in the same file.

Thanks, Effective Typescript!